### PR TITLE
add asynchronous processor and bulk write to DB

### PR DESCRIPTION
**Pull request by danduk82**

### CODE REVIEW

File: sqlalchemylogger/filters.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    def filter(self, record):
+        return bool(re.search(self.regex, record.msg))
+
+
+class DoesNotContainExpression(logging.Filter):
+    """
+    Returns True if the regex is NOT matched in the log message
+    """
+    def __init__(self, regex):
+        self.regex = regex
```
Comment (pvalsecc) [inline sqlalchemylogger/filters.py:21]: You could compile de regex:
```
self.regex = re.compile(regex)
```

And, then, use it like that:
```
self.regex.search(record.msg)
```

File: sqlalchemylogger/handlers.py [hunk 2]
```diff
@@ -1,42 +1,109 @@
...
+                except queue.Empty:
+                     pass
+                if len(logs) > 0:
+                    if (len(logs) >= self.MAX_NB_LOGS) or (time_since_last + self.MAX_TIMEOUT <= time.time()) :
+                        self._writeLogs(logs)
+                        self.log_queue.task_done()
+                        break
+
+
+    def _writeLogs(self,logs):
```
Comment (pvalsecc) [inline sqlalchemylogger/handlers.py:65]: python methods are not camelcased: _write_logs

File: sqlalchemylogger/handlers.py [hunk 3]
```diff
@@ -1,42 +1,109 @@
...
+        _terminated = False
+        while not _terminated:
+            logs = []
+            time_since_last = time.time()
+            while True:
+                try:
+                    log = self.log_queue.get(timeout=self.MAX_TIMEOUT)
+                    if log is None:
+                        # one way of killing a thread
+                        _terminated = True
```
Comment (pvalsecc) [inline sqlalchemylogger/handlers.py:54]: I don't get why you stop the thread when there is no log after 1s?

File: sqlalchemylogger/handlers.py [hunk 4]
```diff
@@ -1,42 +1,109 @@
...
+        self.processor_thread = threading.Thread(target = self._processor, daemon = True)
+        self.processor_thread.start()
+        # initialize filters
+        if doesNotContainExpression:
+             self.addFilter(DoesNotContainExpression(doesNotContainExpression))
+        if containsExpression:
+             self.addFilter(ContainsExpression(containsExpression))
+
+
+    def _processor(self):
```
Comment (pvalsecc) [inline sqlalchemylogger/handlers.py:44]: I would do something like that, instead:
```python
from threading import Condition
import logging


class SQLAlchemyHandler(logging.Handler):
    def __init__(self):
        self.condition = Condition()
        self.queue = []
        # init the rest

    def _processor(self):
        while True:
            with self.condition:
                while not self.queue:
                    self.condition.wait()
                queue = self.queue
                self.queue = []
            self._write_logs(queue)

    def emit(self, record):
        log = self.Log(...)
        with self.condition:
            self.queue.append(log)
            self.condition.notify()
```

File: sqlalchemylogger/handlers.py (hunk 3 above)
Comment (danduk82) [reply sqlalchemylogger/handlers.py:54]: `.get()` does not return a `None`, it returns an exception `Empty`. The way to terminate the thread is to `queue.put(None)` excplicitely

Comment (danduk82) [reply sqlalchemylogger/handlers.py:54]: so the thread is kept alive normally and simply tries to submit logs to `writelogs` each MAX_TIMEOUT

File: sqlalchemylogger/handlers.py (hunk 2 above)
Comment (danduk82) [reply sqlalchemylogger/handlers.py:65]: ok will fix this

File: sqlalchemylogger/filters.py (hunk 1 above)
Comment (danduk82) [reply sqlalchemylogger/filters.py:21]: good idea, thanks

File: sqlalchemylogger/handlers.py (hunk 4 above)
Comment (danduk82) [reply sqlalchemylogger/handlers.py:44]: why do you dislike the usage of proper `queue.Queue()` ? does it add too much overhead?

File: sqlalchemylogger/filters.py (hunk 1 above)
Comment (danduk82) [reply sqlalchemylogger/filters.py:21]: fixed [here](https://github.com/danduk82/poc_oereb_logger/pull/4/commits/e6b198f9257f46965e31d6cb535e389ab68bc1e4)

File: sqlalchemylogger/handlers.py (hunk 2 above)
Comment (danduk82) [reply sqlalchemylogger/handlers.py:65]: fixed [here](https://github.com/danduk82/poc_oereb_logger/pull/4/commits/e6b198f9257f46965e31d6cb535e389ab68bc1e4)

File: sqlalchemylogger/handlers.py (hunk 4 above)
Comment (pvalsecc) [reply sqlalchemylogger/handlers.py:44]: It's more code...
E=mc²
Errors = more_code²

Comment (danduk82) [reply sqlalchemylogger/handlers.py:44]: https://github.com/danduk82/poc_oereb_logger/pull/4/commits/d20a0a2e66e76360e7fba31200f9659dd576d493

**Pull request merged**